# Proyek Analisis Data: E-commerce Public Dataset
- **Nama:** Argo Wahyu Utomo (Arguto)
- **Email:** argo.wahyu.utomo@gmail.com
- **ID Dicoding:** B244051F

## Menentukan Pertanyaan Bisnis

### Penjualan
1. Bagaimana tren penjualan (orders, revenue, unit, ATV, AUR) selama beberapa bulan terakhir?
2. Kategori produk apa saja yang paling tinggi penjualannya (orders, revenue, unit)?
3. Kategori produk apa saja yang paling rendah penjualannya (orders, revenue, unit)?

### Metode Pembayaran
4. Metode pembayaran apa yang paling sering digunakan?
5. Berapa rata-rata nilai transaksi untuk setiap metode pembayaran?
6. Bagaimana distribusi jumlah cicilan untuk pembayaran menggunakan kartu kredit?

### Logistik
7. Seberapa akurat estimasi pengiriman dibandingkan dengan tanggal pengiriman sebenarnya?
8. Berapa nilai minimal, rata-rata, dan maksimal biaya pengiriman per kategori produk?

### Kepuasan Pelanggan
9. Bagaimana tingkat kepuasan pelanggan secara keseluruhan?
10. Pembeli dari kota mana yang memiliki tingkat kepuasan tertinggi?
11. Pembeli dari kota mana yang memiliki tingkat kepuasan terendah?
12. Kategori produk apa yang memiliki tingkat kepuasan pelanggan tertinggi?
13. Bagaimana tingkat kepuasan pelanggan berdasarkan lokasi geografis penjual?

### Lokasi Geografis
13. Bagaimana distribusi pelanggan berdasarkan lokasi geografisnya?
14. Bagaimana distribusi penjual berdasarkan lokasi geografisnya?

## Import Semua Packages/Library yang Digunakan

In [1091]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Data Wrangling

### Gathering Data

#### Memuat file CSV

In [1092]:
dir_ = "dataset/"
file_orders = dir_ + "orders_dataset.csv"
file_products = dir_ + "products_dataset.csv"
file_product_category = dir_ + "product_category_name_translation.csv"
file_items = dir_ + "order_items_dataset.csv"
file_payments = dir_ + "order_payments_dataset.csv"
file_reviews =  dir_ + "order_reviews_dataset.csv"
file_customers = dir_ + "customers_dataset.csv"
file_sellers = dir_ + "sellers_dataset.csv" 
file_geolocation = dir_ + "geolocation_dataset.csv" 

# Load each CSV file into a DataFrame
df_orders = pd.read_csv(file_orders)
df_products = pd.read_csv(file_products)
df_product_category = pd.read_csv(file_product_category)
df_items = pd.read_csv(file_items)
df_payments = pd.read_csv(file_payments)
df_reviews = pd.read_csv(file_reviews)
df_customers = pd.read_csv(file_customers)
df_sellers = pd.read_csv(file_sellers)
df_geolocation = pd.read_csv(file_geolocation)

#### Menampilkan DataFrame

In [1093]:
df_dict = {
    "orders": df_orders,
    "products": df_products,
    "product_category": df_product_category,
    "items": df_items,
    "payments": df_payments,
    "reviews": df_reviews,
    "customers": df_customers,
    "sellers": df_sellers,
    "geolocation": df_geolocation, 
}

top_row_count = 3
for table_name, df in df_dict.items():
    print(f"{table_name.title()} (df_{table_name}) top {top_row_count} rows")
    display(df.head(top_row_count))  # Renders in a rich tabular format

Orders (df_orders) top 3 rows


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00

Products (df_products) top 3 rows


product_id product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5            perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                 artes   
2  96bd76ec8810374ed1b65e291975717f         esporte_lazer   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0

Product_Category (df_product_category) top 3 rows


product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto

Items (df_items) top 3 rows


order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   

   shipping_limit_date  price  freight_value  
0  2017-09-19 09:45:35   58.9          13.29  
1  2017-05-03 11:05:13  239.9          19.93  
2  2018-01-18 14:48:30  199.0          17.87

Payments (df_payments) top 3 rows


order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71

Reviews (df_reviews) top 3 rows


review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   

   review_score review_comment_title review_comment_message  \
0             4                  NaN                    NaN   
1             5                  NaN                    NaN   
2             5                  NaN                    NaN   

  review_creation_date review_answer_timestamp  
0  2018-01-18 00:00:00     2018-01-18 21:46:59  
1  2018-03-10 00:00:00     2018-03-11 03:05:13  
2  2018-02-17 00:00:00     2018-02-18 14:36:24

Customers (df_customers) top 3 rows


customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP

Sellers (df_sellers) top 3 rows


seller_id  seller_zip_code_prefix     seller_city  \
0  3442f8959a84dea7ee197c632cb2df15                   13023        campinas   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844      mogi guacu   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031  rio de janeiro   

  seller_state  
0           SP  
1           SP  
2           RJ

Geolocation (df_geolocation) top 3 rows


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP

#### Insights
- Dataset terdiri dari 9 file csv yang kita muat ke dalam 9 DataFrame, yaitu: df_orders, df_products, df_product_category, df_items, df_payments, df_customers, df_sellers, dan df_geolocation.

### Assessing Data

#### Jumlah Baris & Kolom

In [1094]:
# Create a list to store table information
table_dimension = []

# Iterate over the dictionary to collect data
for table_name, df in df_dict.items():
    table_dimension.append({
        "Table": table_name.title(),
        "DataFrame": f"df_{table_name}",
        "Rows": f"{df.shape[0]:,}",
        "Columns": f"{df.shape[1]:,}"
    })

# Create a DataFrame from the collected data
table_dimension_df = pd.DataFrame(table_dimension)

# Display the DataFrame
display(table_dimension_df)

Table            DataFrame       Rows Columns
0            Orders            df_orders     99,441       8
1          Products          df_products     32,951       9
2  Product_Category  df_product_category         71       2
3             Items             df_items    112,650       7
4          Payments          df_payments    103,886       5
5           Reviews           df_reviews     99,224       7
6         Customers         df_customers     99,441       5
7           Sellers           df_sellers      3,095       4
8       Geolocation       df_geolocation  1,000,163       5

#### Data Type, Missing Values, Unique, Duplicate

In [1095]:
# Create a function to show table info
def show_table_info(table_name_or_df, table_name: str = None):
    """
    Display the information of the DataFrame.
    
    Parameters:
        table_name_or_df (str or DataFrame): The name of the DataFrame (as a string) or the DataFrame itself.
        table_name (str, optional): If table_name_or_df is a DataFrame, provide the variable name here.
    """
    if isinstance(table_name_or_df, str):
        df = df_dict.get(table_name_or_df)
        table_name = table_name_or_df
    elif isinstance(table_name_or_df, pd.DataFrame):
        df = table_name_or_df
        if not table_name:
            table_name = "Unnamed DataFrame"
    else:
        raise ValueError("Input must be either a string (table name) or a pandas DataFrame.")
    
    if df is None:
        raise ValueError(f"No DataFrame found for table name: {table_name_or_df}")
    
    # Initialize a list to collect column information
    table_column_info = []
    
    # Get memory usage for each column
    memory_usage = df.memory_usage(deep=True)
    total_memory_usage = memory_usage.sum()

    # Iterate over each column in the DataFrame
    for column_name in df.columns:
        # Collect information about the column
        table_column_info.append({
            "Table": table_name.title(),
            "Column": column_name,
            "DType": df[column_name].dtype,
            "Not-Null": f"{df[column_name].notnull().sum():,}",
            "Missing Values": f"{df[column_name].isnull().sum():,}",
            "Unique": f"{df[column_name].nunique():,}",
            # "Duplicate": df[column_name].duplicated().sum(),
            "Duplicate": "",  # tidak perlu menghitung duplicated data per kolom
            "Memory Usage": memory_usage[column_name],
        })
    
    # Add a row for total memory usage (aggregate row)
    table_column_info.append({
        "Table": table_name.title(),
        "Column": "[Total]",
        "DType": "",
        "Not-Null": f"{df.notnull().all(axis=1).sum():,}",  # Count rows where all columns are non-null
        "Missing Values": f"{df.isnull().any(axis=1).sum():,}",  # Count rows with any null values
        "Unique": f"{df.shape[0] - df.duplicated().sum():,}",  # Count unique rows without dropping duplicates
        "Duplicate": f"{df.duplicated().sum():,}",  # Count total duplicated rows
        "Memory Usage": total_memory_usage,
    })

    # Create a DataFrame from the collected information
    table_column_info_df = pd.DataFrame(table_column_info)

    # Convert memory usage to human-readable format (e.g., KB, MB) if desired
    table_column_info_df["Memory Usage"] = table_column_info_df["Memory Usage"].apply(
        lambda x: (
            f"{x / (1024**3):.2f} GB" if x >= 1024**3 else
            f"{x / (1024**2):.2f} MB" if x >= 1024**2 else
            f"{x / 1024:.2f} KB" if x >= 1024 else
            f"{x} Bytes"
        )
    )

     # Display the DataFrame
    display(table_column_info_df)


**Catatan**
- Total Not-Null adalah jumlah baris yang seluruh kolomnya tidak mengandung missing value
- Total Missing Value adalah jumlah baris yang mengandung sekurang-kurangnya 1 missing value
- Total Unique adalah jumlah baris unik berdasarkan nilai seluruh kolom
- Total Duplicate adalah jumlah baris yang duplicated berdasarkan nilai dari seluruh kolom
- Tidak perlu menghitung jumlah duplicate pada tiap kolom

##### df_orders

In [1096]:
show_table_info("orders")

Table                         Column   DType Not-Null Missing Values  \
0  Orders                       order_id  object   99,441              0   
1  Orders                    customer_id  object   99,441              0   
2  Orders                   order_status  object   99,441              0   
3  Orders       order_purchase_timestamp  object   99,441              0   
4  Orders              order_approved_at  object   99,281            160   
5  Orders   order_delivered_carrier_date  object   97,658          1,783   
6  Orders  order_delivered_customer_date  object   96,476          2,965   
7  Orders  order_estimated_delivery_date  object   99,441              0   
8  Orders                        [Total]           96,461          2,980   

   Unique Duplicate Memory Usage  
0  99,441                7.68 MB  
1  99,441                7.68 MB  
2       8                5.50 MB  
3  98,875                6.45 MB  
4  90,733                6.44 MB  
5  81,018                6.39 MB  
6  95,664                6.35 MB  
7     459                6.45 MB  
8  99,441         0     52.94 MB

**Insight:**
- Tabel orders (df_orders) memiliki 8 kolom
- tipe data `order_purchase_timestamp`, `order_approved_at`, `order_delivered_carrier_date`, `order_delivered_customer_date`, dan `order_estimated_delivery_date` masih perlu diubah di tahap selanjutnya (data cleaning)
- kolom `order_approved_at` mengandung missing value, kemungkinan besar karena order memang belum di-approve
- kolom `order_delivered_carrier_date` dan `order_delivered_customer_date` juga mengandung missing value, kemungkinan besar karena order belum dikirim atau data delivery memang belum saatnya terisi
- tidak terdapat data duplikat
- terdapat 99,441 pesanan berbeda dengan 8 status berbeda dari 99,441 pelanggan
- dari seluruh pesanan, baru terdapat 90,733 pesanan yang disetujui 

##### df_products

In [1097]:
show_table_info("products")

Table                      Column    DType Not-Null Missing Values  \
0  Products                  product_id   object   32,951              0   
1  Products       product_category_name   object   32,341            610   
2  Products         product_name_lenght  float64   32,341            610   
3  Products  product_description_lenght  float64   32,341            610   
4  Products          product_photos_qty  float64   32,341            610   
5  Products            product_weight_g  float64   32,949              2   
6  Products           product_length_cm  float64   32,949              2   
7  Products           product_height_cm  float64   32,949              2   
8  Products            product_width_cm  float64   32,949              2   
9  Products                     [Total]            32,340            611   

   Unique Duplicate Memory Usage  
0  32,951                2.55 MB  
1      73                1.99 MB  
2      66              257.43 KB  
3   2,960              257.43 KB  
4      19              257.43 KB  
5   2,204              257.43 KB  
6      99              257.43 KB  
7     102              257.43 KB  
8      95              257.43 KB  
9  32,951         0      6.30 MB

**Insight:**
- tabel products (df_products) memiliki 9 kolom
- tipe data semua kolom sudah OK
- banyak terdapat missing values
- tidak terdapat data duplikat
- terdapat 32,951 produk

##### df_product_category

In [1098]:
show_table_info("product_category")

Table                         Column   DType Not-Null  \
0  Product_Category          product_category_name  object       71   
1  Product_Category  product_category_name_english  object       71   
2  Product_Category                        [Total]               71   

  Missing Values Unique Duplicate Memory Usage  
0              0     71                4.57 KB  
1              0     71                4.51 KB  
2              0     71         0      9.21 KB

**Insight:**
- tabel product_category (df_product_category) memiliki 2 kolom
- tipe data seluruh kolom sudah OK
- tidak terdapat missing value
- tidak terdapat data duplikat
- terdapat 71 kategori produk

##### df_items

In [1099]:
show_table_info("items")

Table               Column    DType Not-Null Missing Values   Unique  \
0  Items             order_id   object  112,650              0   98,666   
1  Items        order_item_id    int64  112,650              0       21   
2  Items           product_id   object  112,650              0   32,951   
3  Items            seller_id   object  112,650              0    3,095   
4  Items  shipping_limit_date   object  112,650              0   93,318   
5  Items                price  float64  112,650              0    5,968   
6  Items        freight_value  float64  112,650              0    6,999   
7  Items              [Total]           112,650              0  112,650   

  Duplicate Memory Usage  
0                8.70 MB  
1              880.08 KB  
2                8.70 MB  
3                8.70 MB  
4                7.31 MB  
5              880.08 KB  
6              880.08 KB  
7         0     35.99 MB

In [1100]:
unique_order_item_id = df_items['order_item_id'].unique()
print(unique_order_item_id)

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21]


**Insight:**
- tabel items (df_items) memiliki 7 kolom
- tipe data `order_item_id` dan `shipping_limit_date` perlu diubah
- tidak terdapat missing value
- tidak terdapat data duplikat
- unique value data `order_item_id` kenapa sangat sedikit ya? -> mungkin ID nya hanya unique per order_id, tidak unique terhadap seluruh tabel 

##### df_payments

In [1101]:
show_table_info("payments")

Table                Column    DType Not-Null Missing Values   Unique  \
0  Payments              order_id   object  103,886              0   99,440   
1  Payments    payment_sequential    int64  103,886              0       29   
2  Payments          payment_type   object  103,886              0        5   
3  Payments  payment_installments    int64  103,886              0       24   
4  Payments         payment_value  float64  103,886              0   29,077   
5  Payments               [Total]           103,886              0  103,886   

  Duplicate Memory Usage  
0                8.02 MB  
1              811.61 KB  
2                5.83 MB  
3              811.61 KB  
4              811.61 KB  
5         0     16.23 MB

**Insight:**
- tabel payments (df_payments) memiliki 5 kolom
- tipe data seluruh kolom sudah OK
- tidak terdapat missing values
- tidak terdapat data duplikat
- terdapat 103,886 pembayaran untuk 99,440 order berbeda dengan 5 pilihan tipe pembayaran

##### df_reviews

In [1102]:
show_table_info("reviews")

Table                   Column   DType Not-Null Missing Values  Unique  \
0  Reviews                review_id  object   99,224              0  98,410   
1  Reviews                 order_id  object   99,224              0  98,673   
2  Reviews             review_score   int64   99,224              0       5   
3  Reviews     review_comment_title  object   11,568         87,656   4,527   
4  Reviews   review_comment_message  object   40,977         58,247  36,159   
5  Reviews     review_creation_date  object   99,224              0     636   
6  Reviews  review_answer_timestamp  object   99,224              0  98,248   
7  Reviews                  [Total]            9,839         89,385  99,224   

  Duplicate Memory Usage  
0                7.66 MB  
1                7.66 MB  
2              775.19 KB  
3                3.39 MB  
4                6.78 MB  
5                6.43 MB  
6                6.43 MB  
7         0     39.12 MB

**Insight:**
- tabel reviews (df_reviews) memiliki 7 kolom
- tipe data `review_creation_date` dan `review_answer_timestamp` perlu disesuaikan
- terdapat 89,385 baris dengan missing values
- kolom `review_comment_title` memiliki 87,656 missing values
- kolom `review_comment_message` memiliki 58,247 missing values
- tidak terdapat data duplikat
- terdapat 98,410 review dari 98,673 order berbeda dengan 5 pilihan score

##### df_customers

In [1103]:
show_table_info("customers")

Table                    Column   DType Not-Null Missing Values  \
0  Customers               customer_id  object   99,441              0   
1  Customers        customer_unique_id  object   99,441              0   
2  Customers  customer_zip_code_prefix   int64   99,441              0   
3  Customers             customer_city  object   99,441              0   
4  Customers            customer_state  object   99,441              0   
5  Customers                   [Total]           99,441              0   

   Unique Duplicate Memory Usage  
0  99,441                7.68 MB  
1  96,096                7.68 MB  
2  14,994              776.88 KB  
3   4,119                5.63 MB  
4      27                4.84 MB  
5  99,441         0     26.59 MB

**Insight:**
- tabel customers (df_customerss) memiliki 5 kolom
- tipe data `customer_zip_code_prefix` lebih tepat jika diubah ke string/object
- tidak terdapat missing values
- tidak terdapat data duplikat
- terdapat 99,441 pelanggan dengan 14,994 kode pos berbeda dari 4,119 kota dan 27 state

##### df_sellers

In [1104]:
show_table_info("sellers")

Table                  Column   DType Not-Null Missing Values Unique  \
0  Sellers               seller_id  object    3,095              0  3,095   
1  Sellers  seller_zip_code_prefix   int64    3,095              0  2,246   
2  Sellers             seller_city  object    3,095              0    611   
3  Sellers            seller_state  object    3,095              0     23   
4  Sellers                 [Total]            3,095              0  3,095   

  Duplicate Memory Usage  
0              244.82 KB  
1               24.18 KB  
2              178.89 KB  
3              154.15 KB  
4         0    602.16 KB

**Insight:**
- tabel sellers (df_sellers) memiliki 4 kolom
- tipe data `seller_zip_code_prefix` lebih tepat jika diubah ke string/object
- tidak terdapat missing values
- tidak terdapat data duplikat
- terdapat 3,095 penjual dengan 2,246 kode pos berbeda dari 611 kota dan 23 state

##### df_geolocation

In [1105]:
show_table_info("geolocation")

Table                       Column    DType   Not-Null  \
0  Geolocation  geolocation_zip_code_prefix    int64  1,000,163   
1  Geolocation              geolocation_lat  float64  1,000,163   
2  Geolocation              geolocation_lng  float64  1,000,163   
3  Geolocation             geolocation_city   object  1,000,163   
4  Geolocation            geolocation_state   object  1,000,163   
5  Geolocation                      [Total]           1,000,163   

  Missing Values   Unique Duplicate Memory Usage  
0              0   19,015                7.63 MB  
1              0  717,360                7.63 MB  
2              0  717,613                7.63 MB  
3              0    8,011               57.84 MB  
4              0       27               48.65 MB  
5              0  738,332   261,831    129.38 MB

**Insight:**
- tabel geolocation (df_geolocation) memiliki 5 kolom
- tipe data `geolocation_zip_code_prefix` lebih tepat jika diubah ke string/object
- tidak terdapat missing values
- terdapat data duplikat sebanyak 261,831
- jumlah unique geolocation: 738,332

#### Descriptive Statistics, Potential Outliers

In [1106]:
# Create a function to show table description
def get_table_desc(table_name_or_df, table_name: str = None):
    """
    Display the description of the DataFrame.
    
    Parameters:
        table_name_or_df (str or DataFrame): The name of the DataFrame (as a string) or the DataFrame itself.
        table_name (str, optional): If table_name_or_df is a DataFrame, provide the variable name here.
    """
    if isinstance(table_name_or_df, str):
        df = df_dict.get(table_name_or_df)
        table_name = table_name_or_df
    elif isinstance(table_name_or_df, pd.DataFrame):
        df = table_name_or_df
        if not table_name:
            table_name = "Unnamed DataFrame"
    else:
        raise ValueError("Input must be either a string (table name) or a pandas DataFrame.")
    
    if df is None:
        raise ValueError(f"No DataFrame found for table name: {table_name_or_df}")
    
    # Initialize a list to collect column information
    table_desc = []

    # Iterate over each column in the DataFrame
    for column_name in df.columns:
        # Initialize a dictionary to hold the statistics for each column
        desc = {
            "Table": table_name.title(),
            "Column": column_name,
            "Dtype": str(df[column_name].dtype)
        }
        
        # If the column is numeric, calculate the statistics
        if df[column_name].dtype in ['int64', 'float64']:
            desc["Min"] = f"{df[column_name].min():,.2f}"
            desc["Max"] = f"{df[column_name].max():,.2f}"
            desc["Mean"] = f"{df[column_name].mean():,.2f}"
            desc["Q1"] = f"{df[column_name].quantile(0.25):,.2f}"
            desc["Q2"] = f"{df[column_name].median():,.2f}"
            desc["Q3"] = f"{df[column_name].quantile(0.75):,.2f}"
        elif np.issubdtype(df[column_name].dtype, np.datetime64):
            desc["Min"] = df[column_name].min().strftime("%Y-%m-%d %H:%M:%S")
            desc["Max"] = df[column_name].max().strftime("%Y-%m-%d %H:%M:%S")
        else:
            # For non-numeric columns, only include relevant statistics
            # desc["Mode"] = df[column_name].mode()[0]  # Most frequent value (mode) for non-numeric columns
            desc["Unique"] = f"{df[column_name].nunique():,}"
            desc["Sample Value"] = ', '.join(df[column_name].unique()[:1].astype(str))
        
        # Append the column statistics to the list
        table_desc.append(desc)

    # Create a DataFrame from the collected column descriptions
    table_desc_df = pd.DataFrame(table_desc)
    
    # Remove columns that contain all None values
    columns_to_keep = [col for col in table_desc_df.columns 
                      if not table_desc_df[col].isna().all()]
    
    table_desc_df = table_desc_df.fillna("")
    table_desc_df = table_desc_df[columns_to_keep]

    # Return the DataFrame
    return table_desc_df


def show_table_desc(*args):
    """
    Show the description of a table by displaying the output of get_table_desc.
    
    This function handles both the table name (string) and DataFrame as input.
    """
    display(get_table_desc(*args))


##### df_orders

In [1107]:
show_table_desc("orders")

Table                         Column   Dtype  Unique  \
0  Orders                       order_id  object  99,441   
1  Orders                    customer_id  object  99,441   
2  Orders                   order_status  object       8   
3  Orders       order_purchase_timestamp  object  98,875   
4  Orders              order_approved_at  object  90,733   
5  Orders   order_delivered_carrier_date  object  81,018   
6  Orders  order_delivered_customer_date  object  95,664   
7  Orders  order_estimated_delivery_date  object     459   

                       Sample Value  
0  e481f51cbdc54678b7cc49136f2d6af7  
1  9ef432eb6251297304e76186b10a928d  
2                         delivered  
3               2017-10-02 10:56:33  
4               2017-10-02 11:07:15  
5               2017-10-04 19:55:00  
6               2017-10-10 21:25:13  
7               2017-10-18 00:00:00

**Insights:**
- Perlu mengubah tipe data beberapa kolom untuk bisa menghitung angka statistik deskriptifnya

##### df_products

In [1108]:
show_table_desc("products")

Table                      Column    Dtype  Unique  \
0  Products                  product_id   object  32,951   
1  Products       product_category_name   object      73   
2  Products         product_name_lenght  float64           
3  Products  product_description_lenght  float64           
4  Products          product_photos_qty  float64           
5  Products            product_weight_g  float64           
6  Products           product_length_cm  float64           
7  Products           product_height_cm  float64           
8  Products            product_width_cm  float64           

                       Sample Value   Min        Max      Mean      Q1  \
0  1e9e8ef04dbcff4541ed26657ea517e5                                      
1                        perfumaria                                      
2                                    5.00      76.00     48.48   42.00   
3                                    4.00   3,992.00    771.50  339.00   
4                                    1.00      20.00      2.19    1.00   
5                                    0.00  40,425.00  2,276.47  300.00   
6                                    7.00     105.00     30.82   18.00   
7                                    2.00     105.00     16.94    8.00   
8                                    6.00     118.00     23.20   15.00   

       Q2        Q3  
0                    
1                    
2   51.00     57.00  
3  595.00    972.00  
4    1.00      3.00  
5  700.00  1,900.00  
6   25.00     38.00  
7   13.00     21.00  
8   20.00     30.00

**Insights:**
- Kelihatannya sudah OK, nanti kalau ada kolom yang memang diperlukan untuk analisis baru kita bisa cek lebih lanjut

##### df_product_category

In [1109]:
show_table_desc("product_category")

Table                         Column   Dtype Unique  \
0  Product_Category          product_category_name  object     71   
1  Product_Category  product_category_name_english  object     71   

    Sample Value  
0   beleza_saude  
1  health_beauty

**Insights:**
- Kelihatannya sudah OK

##### df_items

In [1110]:
show_table_desc("items")

Table               Column    Dtype  Unique  \
0  Items             order_id   object  98,666   
1  Items        order_item_id    int64           
2  Items           product_id   object  32,951   
3  Items            seller_id   object   3,095   
4  Items  shipping_limit_date   object  93,318   
5  Items                price  float64           
6  Items        freight_value  float64           

                       Sample Value   Min       Max    Mean     Q1     Q2  \
0  00010242fe8c5a6d1ba2dd792cb16214                                         
1                                    1.00     21.00    1.20   1.00   1.00   
2  4244733e06e7ecb4970a6e2683c13e61                                         
3  48436dade18ac8b2bce089ec2a041202                                         
4               2017-09-19 09:45:35                                         
5                                    0.85  6,735.00  120.65  39.90  74.99   
6                                    0.00    409.68   19.99  13.08  16.26   

       Q3  
0          
1    1.00  
2          
3          
4          
5  134.90  
6   21.15

**Insights:**
- Kita perlu convert tipe data `shipping_limit_date` untuk menghitung statistik deskriptifnya & melihat potensi outlier

##### df_payments

In [1111]:
show_table_desc("payments")

Table                Column    Dtype  Unique  \
0  Payments              order_id   object  99,440   
1  Payments    payment_sequential    int64           
2  Payments          payment_type   object       5   
3  Payments  payment_installments    int64           
4  Payments         payment_value  float64           

                       Sample Value   Min        Max    Mean     Q1      Q2  \
0  b81ef226f3fe1789b1e8b2acac839d17                                           
1                                    1.00      29.00    1.09   1.00    1.00   
2                       credit_card                                           
3                                    0.00      24.00    2.85   1.00    1.00   
4                                    0.00  13,664.08  154.10  56.79  100.00   

       Q3  
0          
1    1.00  
2          
3    4.00  
4  171.84

**Insights:**
- `payment_sequential` perlu dicek lebih lanjut distribusinya, Q1 = Q2 = Q3 = 1, namun nilai Max 29 (potensi outlier, namun belum tentu outlier)
- `payment_installment` perlu dicek lebih lanjut distribusinya (potensi outlier di nilai Max)
- kedua hal di atas untuk sementara bisa diabaikan jika kolom-kolom tersebut belum diperlukan dalam analisis

##### df_reviews

In [1112]:
show_table_desc("reviews")

Table                   Column   Dtype  Unique  \
0  Reviews                review_id  object  98,410   
1  Reviews                 order_id  object  98,673   
2  Reviews             review_score   int64           
3  Reviews     review_comment_title  object   4,527   
4  Reviews   review_comment_message  object  36,159   
5  Reviews     review_creation_date  object     636   
6  Reviews  review_answer_timestamp  object  98,248   

                       Sample Value   Min   Max  Mean    Q1    Q2    Q3  
0  7bc2406110b926393aa56f80a40eba40                                      
1  73fc7af87114b39712e6da79b0a377eb                                      
2                                    1.00  5.00  4.09  4.00  5.00  5.00  
3                               nan                                      
4                               nan                                      
5               2018-01-18 00:00:00                                      
6               2018-01-18 21:46:59

**Insights:**
- kolom `review_creation_date` dan `review_answer_timestamp` perlu diubah dulu tipe datanya untuk menghitung statistik deskriptifnya dan melihat potensi outlier

##### df_customers

In [1113]:
show_table_desc("customers")

Table                    Column   Dtype  Unique  \
0  Customers               customer_id  object  99,441   
1  Customers        customer_unique_id  object  96,096   
2  Customers  customer_zip_code_prefix   int64           
3  Customers             customer_city  object   4,119   
4  Customers            customer_state  object      27   

                       Sample Value       Min        Max       Mean  \
0  06b8999e2fba1a1fbc88172c00ba8bc7                                   
1  861eff4711a542e4b93843c6dd7febb0                                   
2                                    1,003.00  99,990.00  35,137.47   
3                            franca                                   
4                                SP                                   

          Q1         Q2         Q3  
0                                   
1                                   
2  11,347.00  24,416.00  58,900.00  
3                                   
4

**Insights:**
- Selain issue tipe data kolom `customer_zip_code_prefix`, kelihatannya sudah OK

##### df_sellers

In [1114]:
show_table_desc("sellers")

Table                  Column   Dtype Unique  \
0  Sellers               seller_id  object  3,095   
1  Sellers  seller_zip_code_prefix   int64          
2  Sellers             seller_city  object    611   
3  Sellers            seller_state  object     23   

                       Sample Value       Min        Max       Mean        Q1  \
0  3442f8959a84dea7ee197c632cb2df15                                             
1                                    1,001.00  99,730.00  32,291.06  7,093.50   
2                          campinas                                             
3                                SP                                             

          Q2         Q3  
0                        
1  14,940.00  64,552.50  
2                        
3

**Insights:**
- Selain issue tipe data kolom `seller_zip_code_prefix`, kelihatannya sudah OK

##### df_geolocation

In [1115]:
show_table_desc("geolocation")

Table                       Column    Dtype       Min        Max  \
0  Geolocation  geolocation_zip_code_prefix    int64  1,001.00  99,990.00   
1  Geolocation              geolocation_lat  float64    -36.61      45.07   
2  Geolocation              geolocation_lng  float64   -101.47     121.11   
3  Geolocation             geolocation_city   object                        
4  Geolocation            geolocation_state   object                        

        Mean         Q1         Q2         Q3 Unique Sample Value  
0  36,574.17  11,075.00  26,530.00  63,504.00                      
1     -21.18     -23.60     -22.92     -19.98                      
2     -46.39     -48.57     -46.64     -43.77                      
3                                              8,011    sao paulo  
4                                                 27           SP

**Insights:**
- Selain issue tipe data `geolocation_zip_code_prefix`, kelihatannya sudah OK

### Cleaning Data

#### Convert Data Type

##### df_orders

In [1116]:
# List of columns whose data type will be converted 
datetime_columns = [
    'order_purchase_timestamp', 
    'order_approved_at', 
    'order_delivered_carrier_date', 
    'order_delivered_customer_date', 
    'order_estimated_delivery_date'
]

# Convert the specified columns to datetime
for column in datetime_columns:
    df_orders[column] = pd.to_datetime(df_orders[column], errors='coerce')  # 'coerce' will set invalid parsing to NaT

# Verify the change
show_table_desc("orders")

Table                         Column           Dtype  Unique  \
0  Orders                       order_id          object  99,441   
1  Orders                    customer_id          object  99,441   
2  Orders                   order_status          object       8   
3  Orders       order_purchase_timestamp  datetime64[ns]           
4  Orders              order_approved_at  datetime64[ns]           
5  Orders   order_delivered_carrier_date  datetime64[ns]           
6  Orders  order_delivered_customer_date  datetime64[ns]           
7  Orders  order_estimated_delivery_date  datetime64[ns]           

                       Sample Value                  Min                  Max  
0  e481f51cbdc54678b7cc49136f2d6af7                                            
1  9ef432eb6251297304e76186b10a928d                                            
2                         delivered                                            
3                                    2016-09-04 21:15:19  2018-10-17 17:30:18  
4                                    2016-09-15 12:16:38  2018-09-03 17:40:06  
5                                    2016-10-08 10:34:01  2018-09-11 19:48:28  
6                                    2016-10-11 13:46:32  2018-10-17 13:22:46  
7                                    2016-09-30 00:00:00  2018-11-12 00:00:00

**Insight:**
- tipe data kolom `order_purchase_timestamp`, `order_approved_at`, `order_delivered_carrier_date`, `order_delivered_customer_date`, dan `order_estimated_delivery_date` sudah benar

##### df_items

In [1117]:
# List of columns whose data type will be converted 
datetime_columns = [
    'shipping_limit_date', 
]
object_columns = [
    'order_item_id',
]

# Convert the specified columns to datetime
for column in datetime_columns:
    df_items[column] = pd.to_datetime(df_items[column], errors='coerce')  # 'coerce' will set invalid parsing to NaT

# Convert the specified columns to string
for column in object_columns:
    df_items[column] = df_items[column].astype(str)

# Verify the change
show_table_desc("items")

Table               Column           Dtype  Unique  \
0  Items             order_id          object  98,666   
1  Items        order_item_id          object      21   
2  Items           product_id          object  32,951   
3  Items            seller_id          object   3,095   
4  Items  shipping_limit_date  datetime64[ns]           
5  Items                price         float64           
6  Items        freight_value         float64           

                       Sample Value                  Min                  Max  \
0  00010242fe8c5a6d1ba2dd792cb16214                                             
1                                 1                                             
2  4244733e06e7ecb4970a6e2683c13e61                                             
3  48436dade18ac8b2bce089ec2a041202                                             
4                                    2016-09-19 00:15:34  2020-04-09 22:35:08   
5                                                   0.85             6,735.00   
6                                                   0.00               409.68   

     Mean     Q1     Q2      Q3  
0                                
1                                
2                                
3                                
4                                
5  120.65  39.90  74.99  134.90  
6   19.99  13.08  16.26   21.15

**Insight:**
- tipe data kolom `order_item_id` dan `shipping_limit_date` sudah benar

##### df_reviews

In [1118]:
# List of columns whose data type will be converted 
datetime_columns = [
    'review_creation_date',
    'review_answer_timestamp',
]

# Convert the specified columns to datetime
for column in datetime_columns:
    df_reviews[column] = pd.to_datetime(df_reviews[column], errors='coerce')  # 'coerce' will set invalid parsing to NaT

# df_reviews['review_creation_date'] = df_reviews['review_creation_date'].dt.date  # wrong, it will converts to objects

# Verify the change
show_table_desc("reviews")

Table                   Column           Dtype  Unique  \
0  Reviews                review_id          object  98,410   
1  Reviews                 order_id          object  98,673   
2  Reviews             review_score           int64           
3  Reviews     review_comment_title          object   4,527   
4  Reviews   review_comment_message          object  36,159   
5  Reviews     review_creation_date  datetime64[ns]           
6  Reviews  review_answer_timestamp  datetime64[ns]           

                       Sample Value                  Min                  Max  \
0  7bc2406110b926393aa56f80a40eba40                                             
1  73fc7af87114b39712e6da79b0a377eb                                             
2                                                   1.00                 5.00   
3                               nan                                             
4                               nan                                             
5                                    2016-10-02 00:00:00  2018-08-31 00:00:00   
6                                    2016-10-07 18:32:28  2018-10-29 12:27:35   

   Mean    Q1    Q2    Q3  
0                          
1                          
2  4.09  4.00  5.00  5.00  
3                          
4                          
5                          
6

**Insight:**
- tipe data kolom `review_creation_date` dan `review_answer_timestamp` sudah benar

##### df_customers

In [1119]:
# List of columns whose data type will be converted to string
object_columns = [
    'customer_zip_code_prefix',
]

# Convert the specified columns to string
for column in object_columns:
    df_customers[column] = df_customers[column].astype(str)

# Verify the change
show_table_desc("customers")

Table                    Column   Dtype  Unique  \
0  Customers               customer_id  object  99,441   
1  Customers        customer_unique_id  object  96,096   
2  Customers  customer_zip_code_prefix  object  14,994   
3  Customers             customer_city  object   4,119   
4  Customers            customer_state  object      27   

                       Sample Value  
0  06b8999e2fba1a1fbc88172c00ba8bc7  
1  861eff4711a542e4b93843c6dd7febb0  
2                             14409  
3                            franca  
4                                SP

**Insight:**
- tipe data kolom `customer_zip_code_prefix` sudah benar

##### df_sellers

In [1120]:
# List of columns whose data type will be converted to string
object_columns = [
    'seller_zip_code_prefix',
]

# Convert the specified columns to string
for column in object_columns:
    df_sellers[column] = df_sellers[column].astype(str)

# Verify the change
show_table_desc("sellers")

Table                  Column   Dtype Unique  \
0  Sellers               seller_id  object  3,095   
1  Sellers  seller_zip_code_prefix  object  2,246   
2  Sellers             seller_city  object    611   
3  Sellers            seller_state  object     23   

                       Sample Value  
0  3442f8959a84dea7ee197c632cb2df15  
1                             13023  
2                          campinas  
3                                SP

**Insight:**
- tipe data kolom `seller_zip_code_prefix` sudah benar

##### df_geolocation

In [1121]:
# List of columns whose data type will be converted to string
object_columns = [
    'geolocation_zip_code_prefix',
]

# Convert the specified columns to string
for column in object_columns:
    df_geolocation[column] = df_geolocation[column].astype(str)

# Verify the change
show_table_desc("geolocation")

Table                       Column    Dtype  Unique Sample Value  \
0  Geolocation  geolocation_zip_code_prefix   object  19,015         1037   
1  Geolocation              geolocation_lat  float64                        
2  Geolocation              geolocation_lng  float64                        
3  Geolocation             geolocation_city   object   8,011    sao paulo   
4  Geolocation            geolocation_state   object      27           SP   

       Min     Max    Mean      Q1      Q2      Q3  
0                                                   
1   -36.61   45.07  -21.18  -23.60  -22.92  -19.98  
2  -101.47  121.11  -46.39  -48.57  -46.64  -43.77  
3                                                   
4

**Insight:**
- tipe data kolom `geolocation_zip_code_prefix` sudah benar

#### Handle Missing Values

##### df_orders

In [1122]:
# Display rows where 'order_approved_at' is null
df_orders_null_approved = df_orders[df_orders['order_approved_at'].isnull()]
df_orders_null_approved.groupby('order_status')['order_id'].nunique()

order_status
canceled     141
created        5
delivered     14
Name: order_id, dtype: int64

In [1123]:
df_orders.groupby('order_status')['order_id'].nunique()

order_status
approved           2
canceled         625
created            5
delivered      96478
invoiced         314
processing       301
shipped         1107
unavailable      609
Name: order_id, dtype: int64

**Insights**
- Seluruh order dengan status 'created' memiliki missing value pada kolom `order_approved_at`, hal ini wajar karena order dengan status 'created' memang belum disetujui
- Yang menjadi pertanyaan adalah kenapa sebagian order dengan status 'canceled' & 'delivered' juga memiliki missing values di kolom `order_approved_at`?
- Untuk saat ini kita tidak perlu menghapus baris dengan missing value ini karena masih bisa kita asumsikan sebagai catatan order yang valid mengingat seluruh order_id sudah unique (hal ini juga berlaku untuk missing value di kolom `order_delivered_carrier_date` dan `order_delivered_customer_date`)

##### df_products

In [1124]:
df_products['product_category_name'] = df_products['product_category_name'].fillna('[uncategorized]')
show_table_info("products")

Table                      Column    DType Not-Null Missing Values  \
0  Products                  product_id   object   32,951              0   
1  Products       product_category_name   object   32,951              0   
2  Products         product_name_lenght  float64   32,341            610   
3  Products  product_description_lenght  float64   32,341            610   
4  Products          product_photos_qty  float64   32,341            610   
5  Products            product_weight_g  float64   32,949              2   
6  Products           product_length_cm  float64   32,949              2   
7  Products           product_height_cm  float64   32,949              2   
8  Products            product_width_cm  float64   32,949              2   
9  Products                     [Total]            32,340            611   

   Unique Duplicate Memory Usage  
0  32,951                2.55 MB  
1      74                2.01 MB  
2      66              257.43 KB  
3   2,960              257.43 KB  
4      19              257.43 KB  
5   2,204              257.43 KB  
6      99              257.43 KB  
7     102              257.43 KB  
8      95              257.43 KB  
9  32,951         0      6.32 MB

**Insights**
- Missing values di product_category_name sebanyak 610 baris kita isi dengan value '[uncategorized]', kategori baru ini selanjutnya perlu kita tambahkan ke tabel product_category supaya dapat pasangan `product_category_name_english`nya
- Jumlah unique value untuk product_category_name sebanyak 74, sedangkan di tabel product_category hanya ada 72, sehingga perlu kita tambahkan kedua kategori ini ke tabel product_category supaya dapat pasangan `product_category_name_english`nya juga
- Missing values di kolom yang lain untuk saat ini bisa kita abaikan karena tidak relevan dengan tujuan analisis saat ini

##### df_product_category

In [1125]:
df_product_category.tail()

product_category_name product_category_name_english
66                         flores                       flowers
67             artes_e_artesanato         arts_and_craftmanship
68                fraldas_higiene           diapers_and_hygiene
69  fashion_roupa_infanto_juvenil     fashion_childrens_clothes
70             seguros_e_servicos         security_and_services

In [1126]:
show_table_info("product_category")

Table                         Column   DType Not-Null  \
0  Product_Category          product_category_name  object       71   
1  Product_Category  product_category_name_english  object       71   
2  Product_Category                        [Total]               71   

  Missing Values Unique Duplicate Memory Usage  
0              0     71                4.57 KB  
1              0     71                4.51 KB  
2              0     71         0      9.21 KB

In [1127]:
# Create a new row as a dictionary
new_row = {'product_category_name': '[uncategorized]', 'product_category_name_english': '[uncategorized]'}

# Convert the single row dictionary to a DataFrame
new_row_df = pd.DataFrame([new_row])

# Append the new row to the DataFrame
df_product_category = pd.concat([df_product_category, new_row_df], ignore_index=True)

df_product_category.tail()

product_category_name product_category_name_english
67             artes_e_artesanato         arts_and_craftmanship
68                fraldas_higiene           diapers_and_hygiene
69  fashion_roupa_infanto_juvenil     fashion_childrens_clothes
70             seguros_e_servicos         security_and_services
71                [uncategorized]               [uncategorized]

In [1128]:
# Update the df_dict to reflect the changes
# It is needed because adding a row creates a new DataFrame,
# which requires updating the reference in df_dict.
df_dict["product_category"] = df_product_category

In [1129]:
show_table_info("product_category")

Table                         Column   DType Not-Null  \
0  Product_Category          product_category_name  object       72   
1  Product_Category  product_category_name_english  object       72   
2  Product_Category                        [Total]               72   

  Missing Values Unique Duplicate Memory Usage  
0              0     72                4.64 KB  
1              0     72                4.57 KB  
2              0     72         0      9.34 KB

**Insight:**
- kita sudah menambahkan satu kategori baru `[uncategorized]` ke dalam tabel product_category (df_product_category) supaya terintegrasi dengan perubahan pada tabel sebelumnya yaitu tabel produk (df_products)

In [1130]:
# Get set of product_category_name from df_products
product_category_from_products = set(df_products['product_category_name'].unique())

# # Get set of product_category_name from df_product_category
product_category = set(df_product_category['product_category_name'].unique())

# # Find the difference (product categories in df_orders but not in df_product_category)
df_product_category_no_english = product_category_from_products.difference(product_category)
df_product_category_no_english

{'pc_gamer', 'portateis_cozinha_e_preparadores_de_alimentos'}

In [1131]:
# Let's put the English translation for those two category into df_product_category
# Define the new rows to be added with both category name and English translation
new_rows = [
    {'product_category_name': 'pc_gamer', 'product_category_name_english': 'pc_gamer'},
    {'product_category_name': 'portateis_cozinha_e_preparadores_de_alimentos', 'product_category_name_english': 'portable_kitchen_and_food_preparers'}
]

# Convert the rows to a DataFrame
new_rows_df = pd.DataFrame(new_rows)

# Append the new rows to the existing df_product_category DataFrame
df_product_category = pd.concat([df_product_category, new_rows_df], ignore_index=True)

# Reassign the updated DataFrame back to the df_dict
df_dict["product_category"] = df_product_category

# Verify the changes
show_table_info("product_category")

Table                         Column   DType Not-Null  \
0  Product_Category          product_category_name  object       74   
1  Product_Category  product_category_name_english  object       74   
2  Product_Category                        [Total]               74   

  Missing Values Unique Duplicate Memory Usage  
0              0     74                4.78 KB  
1              0     74                4.71 KB  
2              0     74         0      9.62 KB

In [1132]:
df_product_category.tail()

product_category_name  \
69                  fashion_roupa_infanto_juvenil   
70                             seguros_e_servicos   
71                                [uncategorized]   
72                                       pc_gamer   
73  portateis_cozinha_e_preparadores_de_alimentos   

          product_category_name_english  
69            fashion_childrens_clothes  
70                security_and_services  
71                      [uncategorized]  
72                             pc_gamer  
73  portable_kitchen_and_food_preparers

**Insight:**
- kita sudah menambahkan dua kategori baru lagi ke dalam tabel product_category (df_product_category) supaya terintegrasi dengan perubahan pada tabel sebelumnya yaitu tabel produk (df_products)

##### df_reviews

**Insight:**
- missing values pada kolom `review_comment_title` dan `review_comment_message` merupakan hal yang wajar dan untuk saat ini tidak perlu kita isi

#### Handle Duplicated Data

##### df_geolocation

In [1133]:
df_geolocation.drop_duplicates(inplace=True)
show_table_info("geolocation")

Table                       Column    DType Not-Null Missing Values  \
0  Geolocation  geolocation_zip_code_prefix   object  738,332              0   
1  Geolocation              geolocation_lat  float64  738,332              0   
2  Geolocation              geolocation_lng  float64  738,332              0   
3  Geolocation             geolocation_city   object  738,332              0   
4  Geolocation            geolocation_state   object  738,332              0   
5  Geolocation                      [Total]           738,332              0   

    Unique Duplicate Memory Usage  
0   19,015               37.87 MB  
1  717,360                5.63 MB  
2  717,613                5.63 MB  
3    8,011               43.59 MB  
4       27               35.91 MB  
5  738,332         0    134.28 MB

**Insight:**
- duplicated data di tabel geolocation sudah tidak ada

## Exploratory Data Analysis (EDA)

### Analisis Penjualan

#### 1. Tren orders, revenue, unit, ATV, AUR

In [1134]:
valid_columns_orders = [
    "order_id",
    "customer_id",
    "order_status",
    "order_purchase_timestamp",
]

valid_columns_items = [
    "order_id",
    "order_item_id",
    "product_id",
    "seller_id",
    "price",
]

# Filter for delivered orders
df_orders_delivered = df_orders[df_orders['order_status'] == "delivered"]

# Merge with order items to include price and other details
df_orders_delivered_items = pd.merge(
    df_orders_delivered[valid_columns_orders], 
    df_items[valid_columns_items], 
    on='order_id', 
    how='left'
)

show_table_info(df_orders_delivered_items, "delivered_orders")

Table                    Column           DType Not-Null  \
0  Delivered_Orders                  order_id          object  110,197   
1  Delivered_Orders               customer_id          object  110,197   
2  Delivered_Orders              order_status          object  110,197   
3  Delivered_Orders  order_purchase_timestamp  datetime64[ns]  110,197   
4  Delivered_Orders             order_item_id          object  110,197   
5  Delivered_Orders                product_id          object  110,197   
6  Delivered_Orders                 seller_id          object  110,197   
7  Delivered_Orders                     price         float64  110,197   
8  Delivered_Orders                   [Total]                  110,197   

  Missing Values   Unique Duplicate Memory Usage  
0              0   96,478                8.51 MB  
1              0   96,478                8.51 MB  
2              0        1                6.10 MB  
3              0   95,956              860.91 KB  
4              0       21                5.25 MB  
5              0   32,216                8.51 MB  
6              0    2,970                8.51 MB  
7              0    5,859              860.91 KB  
8              0  110,197         0     47.08 MB

In [1135]:
# Group by order_id and calculate 'value' (sum of price) and 'unit' (count of items)
df_orders_sales = df_orders_delivered_items.groupby('order_id').agg(
    value=('price', 'sum'),   # Total price per order
    unit=('order_item_id', 'count')  # Total item count per order
).reset_index()

show_table_info(df_orders_sales, "order_sales")

Table    Column    DType Not-Null Missing Values  Unique Duplicate  \
0  Order_Sales  order_id   object   96,478              0  96,478             
1  Order_Sales     value  float64   96,478              0   7,636             
2  Order_Sales      unit    int64   96,478              0      17             
3  Order_Sales   [Total]            96,478              0  96,478         0   

  Memory Usage  
0      7.45 MB  
1    753.73 KB  
2    753.73 KB  
3      8.92 MB

In [1136]:
# Merge the aggregated data back with the original delivered orders DataFrame
df_orders_delivered_with_sales = pd.merge(
    df_orders_delivered[valid_columns_orders], 
    df_orders_sales, 
    on='order_id', 
    how='inner'
)

# Display the final DataFrame
df_orders_delivered_with_sales.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   value  unit  
0    delivered      2017-10-02 10:56:33   29.99     1  
1    delivered      2018-07-24 20:41:37  118.70     1  
2    delivered      2018-08-08 08:38:49  159.90     1  
3    delivered      2017-11-18 19:28:06   45.00     1  
4    delivered      2018-02-13 21:18:39   19.90     1

In [1137]:
show_table_desc(df_orders_delivered_with_sales, "order_delivered_with_sales")

Table                    Column           Dtype  \
0  Order_Delivered_With_Sales                  order_id          object   
1  Order_Delivered_With_Sales               customer_id          object   
2  Order_Delivered_With_Sales              order_status          object   
3  Order_Delivered_With_Sales  order_purchase_timestamp  datetime64[ns]   
4  Order_Delivered_With_Sales                     value         float64   
5  Order_Delivered_With_Sales                      unit           int64   

   Unique                      Sample Value                  Min  \
0  96,478  e481f51cbdc54678b7cc49136f2d6af7                        
1  96,478  9ef432eb6251297304e76186b10a928d                        
2       1                         delivered                        
3                                            2016-09-15 12:16:38   
4                                                           0.85   
5                                                           1.00   

                   Max    Mean     Q1     Q2      Q3  
0                                                     
1                                                     
2                                                     
3  2018-08-29 15:00:37                                
4            13,440.00  137.04  45.90  86.57  149.90  
5                21.00    1.14   1.00   1.00    1.00

In [1138]:
# Add 'year' and 'month' columns to the DataFrame
df_orders_delivered_with_sales['year'] = df_orders_delivered_with_sales['order_purchase_timestamp'].dt.year
df_orders_delivered_with_sales['month'] = df_orders_delivered_with_sales['order_purchase_timestamp'].dt.month

# Group by 'year' and 'month'
df_orders_monthly = df_orders_delivered_with_sales.groupby(['year', 'month']).agg(
    orders=('order_id', 'nunique'),
    revenue=('value', 'sum'),
    unit=('unit', 'sum'),
).reset_index()

# Calculate ATV (Average Transaction Value) and AUR (Average Unit Revenue)
df_orders_monthly['atv'] = df_orders_monthly['revenue'] / df_orders_monthly['orders']
df_orders_monthly['aur'] = df_orders_monthly['revenue'] / df_orders_monthly['unit']

# Format the result
df_orders_monthly = df_orders_monthly.rename(columns={
    'year': 'Year',
    'month': 'Month',
    'orders': 'Orders',
    'revenue': 'Revenue',
    'unit': 'Unit',
    'atv': 'ATV',
    'aur': 'AUR'
})

# Display the resulting DataFrame
df_orders_monthly


Year  Month  Orders    Revenue  Unit         ATV         AUR
0   2016      9       1     134.97     3  134.970000   44.990000
1   2016     10     265   40325.11   313  152.170226  128.834217
2   2016     12       1      10.90     1   10.900000   10.900000
3   2017      1     750  111798.36   913  149.064480  122.451654
4   2017      2    1653  234223.40  1858  141.695947  126.062110
5   2017      3    2546  359198.85  2897  141.083602  123.989938
6   2017      4    2303  340669.68  2569  147.924307  132.607894
7   2017      5    3546  489338.25  4004  137.997250  122.212350
8   2017      6    3135  421923.37  3489  134.584807  120.929599
9   2017      7    3872  481604.52  4416  124.381333  109.058995
10  2017      8    4193  554699.70  4797  132.291844  115.634709
11  2017      9    4150  607399.67  4737  146.361366  128.224545
12  2017     10    4478  648247.65  5214  144.762762  124.328280
13  2017     11    7289  987765.37  8475  135.514525  116.550486
14  2017     12    5513  726033.19  6187  131.694756  117.348180
15  2018      1    7069  924645.00  8037  130.802801  115.048526
16  2018      2    6555  826437.13  7518  126.077365  109.927791
17  2018      3    7003  953356.25  8017  136.135406  118.916833
18  2018      4    6798  973534.09  7827  143.208898  124.381511
19  2018      5    6749  977544.69  7810  144.842894  125.165773
20  2018      6    6099  856077.86  7010  140.363643  122.122377
21  2018      7    6159  867953.46  6963  140.924413  124.652227
22  2018      8    6351  838576.64  7142  132.038520  117.414819

In [1139]:
# Generate descriptive statistics
desc = df_orders_monthly.describe()

# Format all numeric values to 2 decimal places with a thousand separator
formatted_desc = desc.map(lambda x: f"{x:,.2f}" if isinstance(x, (int, float)) else x)

# Display the formatted DataFrame
display(formatted_desc)

Year  Month    Orders     Revenue      Unit     ATV     AUR
count     23.00  23.00     23.00       23.00     23.00   23.00   23.00
mean   2,017.22   6.30  4,194.70  574,847.74  4,791.17  133.03  113.12
std        0.67   3.47  2,481.18  337,263.53  2,845.46   27.60   27.95
min    2,016.00   1.00      1.00       10.90      1.00   10.90   10.90
25%    2,017.00   3.50  2,424.50  349,934.27  2,733.00  132.17  116.09
50%    2,017.00   6.00  4,193.00  607,399.67  4,797.00  138.00  122.12
75%    2,018.00   9.00  6,453.00  862,015.66  7,330.00  143.99  124.52
max    2,018.00  12.00  7,289.00  987,765.37  8,475.00  152.17  132.61

**Insight:**
- Performa penjualan secara umum cenderung meningkat dari September 2016 sampai Agustus 2018
- Hanya terdapat 1 order pada bulan September dan Desember 2016
- Tidak ada penjualan sama sekali pada bulan November 2016
- Orders tertinggi sebesar 7,003 terjadi pada bulan Maret 2018
- Revenue & Unit Sold tertinggi sebesar 987,765.37 (Real Brazil) & 8,475 unit terjadi pada bulan November 2017
- ATV tertingi sebesar 152.17 terjadi pada Oktober 2016
- AUR tertinggi sebesar 132.61 terjadi pada April 2017

#### 2. Kategori produk terbaik (orders, revenue, unit)

In [1140]:
# Merge products to product category in English
df_product_to_category = pd.merge(
    df_products[['product_id', 'product_category_name']],
    df_product_category,
    on='product_category_name',
    how='left'
)

df_product_to_category.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

  product_category_name_english  
0                     perfumery  
1                           art  
2                sports_leisure  
3                          baby  
4                    housewares

In [1141]:
df_orders_delivered_items.head(3)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp order_item_id  \
0    delivered      2017-10-02 10:56:33             1   
1    delivered      2018-07-24 20:41:37             1   
2    delivered      2018-08-08 08:38:49             1   

                         product_id                         seller_id   price  
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   29.99  
1  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962  118.70  
2  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2  159.90

In [1142]:
df_orders_delivered_product_category = pd.merge(
    df_orders_delivered_items,
    df_product_to_category,
    on='product_id',
    how='left'
)
show_table_desc(df_orders_delivered_product_category, "orders_delivered_product_category")

Table                         Column  \
0  Orders_Delivered_Product_Category                       order_id   
1  Orders_Delivered_Product_Category                    customer_id   
2  Orders_Delivered_Product_Category                   order_status   
3  Orders_Delivered_Product_Category       order_purchase_timestamp   
4  Orders_Delivered_Product_Category                  order_item_id   
5  Orders_Delivered_Product_Category                     product_id   
6  Orders_Delivered_Product_Category                      seller_id   
7  Orders_Delivered_Product_Category                          price   
8  Orders_Delivered_Product_Category          product_category_name   
9  Orders_Delivered_Product_Category  product_category_name_english   

            Dtype  Unique                      Sample Value  \
0          object  96,478  e481f51cbdc54678b7cc49136f2d6af7   
1          object  96,478  9ef432eb6251297304e76186b10a928d   
2          object       1                         delivered   
3  datetime64[ns]                                             
4          object      21                                 1   
5          object  32,216  87285b34884572647811a353c7ac498a   
6          object   2,970  3504c0cb71d7fa48d967e0e4c94d59d9   
7         float64                                             
8          object      74             utilidades_domesticas   
9          object      74                        housewares   

                   Min                  Max    Mean     Q1     Q2      Q3  
0                                                                          
1                                                                          
2                                                                          
3  2016-09-15 12:16:38  2018-08-29 15:00:37                                
4                                                                          
5                                                                          
6                                                                          
7                 0.85             6,735.00  119.98  39.90  74.90  134.17  
8                                                                          
9

In [1143]:
# Group by product_id and calculate 'orders', 'revenue', and 'unit'
df_orders_sales_by_product_category = df_orders_delivered_product_category.groupby('product_category_name_english').agg(
    orders=('order_id', 'count'),
    revenue=('price', 'sum'),
    unit=('order_item_id', 'count')
).reset_index()

# Get sorted category by 'orders' and add rank
df_orders_sales_by_product_category_sorted_orders = (
    df_orders_sales_by_product_category
    .sort_values(by='orders', ascending=False)
    .reset_index(drop=True)
)
df_orders_sales_by_product_category_sorted_orders['orders_rank'] = range(1, len(df_orders_sales_by_product_category_sorted_orders) + 1)

# Get sorted category by 'revenue' and add rank
df_orders_sales_by_product_category_sorted_revenue = (
    df_orders_sales_by_product_category
    .sort_values(by='revenue', ascending=False)
    .reset_index(drop=True)
)
df_orders_sales_by_product_category_sorted_revenue['revenue_rank'] = range(1, len(df_orders_sales_by_product_category_sorted_revenue) + 1)

# Get sorted category by 'unit' and add rank
df_orders_sales_by_product_category_sorted_unit = (
    df_orders_sales_by_product_category
    .sort_values(by='unit', ascending=False)
    .reset_index(drop=True)
)
df_orders_sales_by_product_category_sorted_unit['unit_rank'] = range(1, len(df_orders_sales_by_product_category_sorted_unit) + 1)

desc_top_orders = "\n[Kategori Produk yang Paling Banyak Dipesan (Orders)]"
desc_top_revenue = "\n[Kategori Produk yang Paling Banyak Menghasilkan Revenue)]"
desc_top_unit = "\n[Kategori Produk yang Paling Banyak Terjual (Unit)]"

top_category_dict = {
    "Top Product Categories by Orders" + desc_top_orders: df_orders_sales_by_product_category_sorted_orders,
    "Top Product Categories by Revenue" + desc_top_revenue: df_orders_sales_by_product_category_sorted_revenue,
    "Top Product Categories by Unit" + desc_top_unit: df_orders_sales_by_product_category_sorted_unit,
}

# Format the columns for 'orders', 'revenue' and 'unit' with 2 decimal places and thousands separator
for title, df in top_category_dict.items():
    df['orders'] = df['orders'].apply(lambda x: f"{x:,.2f}")
    df['revenue'] = df['revenue'].apply(lambda x: f"{x:,.2f}")
    df['unit'] = df['unit'].apply(lambda x: f"{x:,.2f}")

    # Display the Results
    print(title)
    display(df.head(10))


Top Product Categories by Orders
[Kategori Produk yang Paling Banyak Dipesan (Orders)]


product_category_name_english     orders       revenue       unit  \
0                bed_bath_table  10,953.00  1,023,434.76  10,953.00   
1                 health_beauty   9,465.00  1,233,131.72   9,465.00   
2                sports_leisure   8,431.00    954,852.55   8,431.00   
3               furniture_decor   8,160.00    711,927.69   8,160.00   
4         computers_accessories   7,644.00    888,724.61   7,644.00   
5                    housewares   6,795.00    615,628.69   6,795.00   
6                 watches_gifts   5,859.00  1,166,176.98   5,859.00   
7                     telephony   4,430.00    309,860.23   4,430.00   
8                  garden_tools   4,268.00    470,495.28   4,268.00   
9                          auto   4,140.00    578,966.65   4,140.00   

   orders_rank  
0            1  
1            2  
2            3  
3            4  
4            5  
5            6  
6            7  
7            8  
8            9  
9           10

Top Product Categories by Revenue
[Kategori Produk yang Paling Banyak Menghasilkan Revenue)]


product_category_name_english     orders       revenue       unit  \
0                 health_beauty   9,465.00  1,233,131.72   9,465.00   
1                 watches_gifts   5,859.00  1,166,176.98   5,859.00   
2                bed_bath_table  10,953.00  1,023,434.76  10,953.00   
3                sports_leisure   8,431.00    954,852.55   8,431.00   
4         computers_accessories   7,644.00    888,724.61   7,644.00   
5               furniture_decor   8,160.00    711,927.69   8,160.00   
6                    housewares   6,795.00    615,628.69   6,795.00   
7                    cool_stuff   3,718.00    610,204.10   3,718.00   
8                          auto   4,140.00    578,966.65   4,140.00   
9                          toys   4,030.00    471,286.48   4,030.00   

   revenue_rank  
0             1  
1             2  
2             3  
3             4  
4             5  
5             6  
6             7  
7             8  
8             9  
9            10

Top Product Categories by Unit
[Kategori Produk yang Paling Banyak Terjual (Unit)]


product_category_name_english     orders       revenue       unit  unit_rank
0                bed_bath_table  10,953.00  1,023,434.76  10,953.00          1
1                 health_beauty   9,465.00  1,233,131.72   9,465.00          2
2                sports_leisure   8,431.00    954,852.55   8,431.00          3
3               furniture_decor   8,160.00    711,927.69   8,160.00          4
4         computers_accessories   7,644.00    888,724.61   7,644.00          5
5                    housewares   6,795.00    615,628.69   6,795.00          6
6                 watches_gifts   5,859.00  1,166,176.98   5,859.00          7
7                     telephony   4,430.00    309,860.23   4,430.00          8
8                  garden_tools   4,268.00    470,495.28   4,268.00          9
9                          auto   4,140.00    578,966.65   4,140.00         10

#### 3. Kategori produk terburuk (orders, revenue, unit)

In [1144]:
# Get sorted category by 'orders' and add rank
df_orders_sales_by_product_category_sorted_orders = (
    df_orders_sales_by_product_category
    .sort_values(by='orders', ascending=True)
    .reset_index(drop=True)
)
df_orders_sales_by_product_category_sorted_orders['orders_rank'] = range(
    len(df_orders_sales_by_product_category_sorted_orders), 0, -1
)

# Get sorted category by 'revenue' and add rank
df_orders_sales_by_product_category_sorted_revenue = (
    df_orders_sales_by_product_category
    .sort_values(by='revenue', ascending=True)
    .reset_index(drop=True)
)
df_orders_sales_by_product_category_sorted_revenue['revenue_rank'] = range(
    len(df_orders_sales_by_product_category_sorted_revenue), 0, -1
)

# Get sorted category by 'unit' and add rank
df_orders_sales_by_product_category_sorted_unit = (
    df_orders_sales_by_product_category
    .sort_values(by='unit', ascending=True)
    .reset_index(drop=True)
)
df_orders_sales_by_product_category_sorted_unit['unit_rank'] = range(
    len(df_orders_sales_by_product_category_sorted_unit), 0, -1
)

desc_bottom_orders = "\n[Kategori Produk yang Paling Sedikit Dipesan (Orders)]"
desc_bottom_revenue = "\n[Kategori Produk yang Paling Sedikit Menghasilkan Revenue)]"
desc_bottom_unit = "\n[Kategori Produk yang Paling Sedikit Terjual (Unit)]"

bottom_category_dict = {
    "Product Categories with the Fewest Orders" + desc_bottom_orders: df_orders_sales_by_product_category_sorted_orders,
    "Product Categories with the Lowest Revenue" + desc_bottom_revenue: df_orders_sales_by_product_category_sorted_revenue,
    "Product Categories with the Fewest Units Sold" + desc_bottom_unit: df_orders_sales_by_product_category_sorted_unit,
}

# Format the columns for 'orders', 'revenue' and 'unit' with 2 decimal places and thousands separator
for title, df in bottom_category_dict.items():
    df['orders'] = df['orders'].apply(lambda x: f"{x:,.2f}")
    df['revenue'] = df['revenue'].apply(lambda x: f"{x:,.2f}")
    df['unit'] = df['unit'].apply(lambda x: f"{x:,.2f}")

    # Display the Results
    print(title)
    display(df.head(10))

Product Categories with the Fewest Orders
[Kategori Produk yang Paling Sedikit Dipesan (Orders)]


product_category_name_english orders   revenue   unit  orders_rank
0                security_and_services   2.00    283.29   2.00           74
1            fashion_childrens_clothes   7.00    519.95   7.00           73
2                             pc_gamer   8.00  1,306.95   8.00           72
3                           la_cuisine  14.00  2,054.99  14.00           71
4  portable_kitchen_and_food_preparers  14.00  3,933.63  14.00           70
5                    cds_dvds_musicals  14.00    730.00  14.00           69
6                arts_and_craftmanship  24.00  1,814.01  24.00           68
7                        fashion_sport  29.00  2,094.52  29.00           67
8                       home_comfort_2  30.00    760.27  30.00           66
9                              flowers  33.00  1,110.04  33.00           65

Product Categories with the Lowest Revenue
[Kategori Produk yang Paling Sedikit Menghasilkan Revenue)]


product_category_name_english orders   revenue   unit  revenue_rank
0         security_and_services   2.00    283.29   2.00            74
1     fashion_childrens_clothes   7.00    519.95   7.00            73
2             cds_dvds_musicals  14.00    730.00  14.00            72
3                home_comfort_2  30.00    760.27  30.00            71
4                       flowers  33.00  1,110.04  33.00            70
5                      pc_gamer   8.00  1,306.95   8.00            69
6           diapers_and_hygiene  37.00  1,500.79  37.00            68
7         arts_and_craftmanship  24.00  1,814.01  24.00            67
8                    la_cuisine  14.00  2,054.99  14.00            66
9                 fashion_sport  29.00  2,094.52  29.00            65

Product Categories with the Fewest Units Sold
[Kategori Produk yang Paling Sedikit Terjual (Unit)]


product_category_name_english orders   revenue   unit  unit_rank
0                security_and_services   2.00    283.29   2.00         74
1            fashion_childrens_clothes   7.00    519.95   7.00         73
2                             pc_gamer   8.00  1,306.95   8.00         72
3                           la_cuisine  14.00  2,054.99  14.00         71
4  portable_kitchen_and_food_preparers  14.00  3,933.63  14.00         70
5                    cds_dvds_musicals  14.00    730.00  14.00         69
6                arts_and_craftmanship  24.00  1,814.01  24.00         68
7                        fashion_sport  29.00  2,094.52  29.00         67
8                       home_comfort_2  30.00    760.27  30.00         66
9                              flowers  33.00  1,110.04  33.00         65

## Visualization & Explanatory Analysis

### Pertanyaan 1:

### Pertanyaan 2:

**Insight:**
- xxx
- xxx

## Analisis Lanjutan (Opsional)

## Conclusion

- Conclution pertanyaan 1
- Conclution pertanyaan 2